File: 백준/Gold/1194. 달이 차오른다， 가자．/달이 차오른다， 가자．.cc
#include <iostream>
#include <algorithm>
#include <vector>
#include <queue>
#include <utility>
#include <string>
#include <cstring>
#include <cstdlib>

using namespace std;

int n, m, ans=-1;
string s[55];
bool visited[55][55][65];

typedef struct MAP {
	int x;
	int y;
	int key;
	int cnt;
}map;

int dx[4] = { -1,1,0,0 };
int dy[4] = { 0,0,-1,1 };

queue<MAP> q;

void dfs(int x, int y, int key, int cnt) {
	if (x < 0 || x >= n || y < 0 || y >= m || visited[x][y][key]) return;

	visited[x][y][key] = true;

	if (s[x][y] == '1') {//탈출
		ans = cnt;
		while (!q.empty()) q.pop();
		return;
	}
	if (s[x][y] == '#') return;
	else if (s[x][y] >= 'a' && s[x][y] <= 'f') {
		key |= 1 << (s[x][y] - 'a');
		//cout << "a : " << key << "\n";
		for (int i = 0; i < 4; i++) {
			map tmp = { x + dx[i],y + dy[i],key,cnt + 1 };
			q.push(tmp);
		}
	}
	else if (s[x][y] >= 'A' && s[x][y] <= 'F') {
		//cout << (key & (1 << (s[x][y] - 'A'))) << '\n';
		if ((key & (1 << (s[x][y] - 'A'))) != 0) {
			for (int i = 0; i < 4; i++) {
				map tmp = { x + dx[i],y + dy[i],key,cnt + 1 };
					q.push(tmp);
			}
		}
	}
	else{
		for (int i = 0; i < 4; i++) {
			map tmp = { x + dx[i],y + dy[i],key,cnt + 1 };
			q.push(tmp);
		}
	}
}

int main() {
	
	cin >> n >> m;
	for (int i = 0; i < n; i++) {
		cin >> s[i];
		for (int j = 0; j < m; j++) {
			if (s[i][j] == '0') {
				map tmp = { i,j,0,0 };
				q.push(tmp);
			}
		}
	}
	while (!q.empty()) {
		map tmp = q.front();
		q.pop();
		dfs(tmp.x, tmp.y, tmp.key, tmp.cnt);
	}
	cout << ans;

	return 0;
}

File: 백준/Gold/1194. 달이 차오른다， 가자．/달이 차오른다， 가자．.java
import java.util.*;

public class Main {
    static int n, m, ans = -1;
    static String[] s;
    static boolean[][][] visited;

    static class MAP {
        int x, y, key, cnt;

        MAP(int x, int y, int key, int cnt) {
            this.x = x;
            this.y = y;
            this.key = key;
            this.cnt = cnt;
        }
    }

    static final int[] dx = {-1, 1, 0, 0};
    static final int[] dy = {0, 0, -1, 1};

    public static void bfs(int x, int y, int key, int cnt, Queue<MAP> q) {
        if (x < 0 || x >= n || y < 0 || y >= m || visited[x][y][key]) return;

        visited[x][y][key] = true;

        if (s[x].charAt(y) == '1') { // Exit
            ans = cnt;
            q.clear(); // Clear the queue
            return;
        }
        if (s[x].charAt(y) == '#') return;
        else if (s[x].charAt(y) >= 'a' && s[x].charAt(y) <= 'f') {
            key |= 1 << (s[x].charAt(y) - 'a');
            for (int i = 0; i < 4; i++) {
                q.add(new MAP(x + dx[i], y + dy[i], key, cnt + 1));
            }
        } else if (s[x].charAt(y) >= 'A' && s[x].charAt(y) <= 'F') {
            if ((key & (1 << (s[x].charAt(y) - 'A'))) != 0) {
                for (int i = 0; i < 4; i++) {
                    q.add(new MAP(x + dx[i], y + dy[i], key, cnt + 1));
                }
            }
        } else {
            for (int i = 0; i < 4; i++) {
                q.add(new MAP(x + dx[i], y + dy[i], key, cnt + 1));
            }
        }
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        n = sc.nextInt();
        m = sc.nextInt();
        s = new String[n];
        visited = new boolean[n][m][65];

        for (int i = 0; i < n; i++) {
            s[i] = sc.next();
        }

        Queue<MAP> q = new LinkedList<>();

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (s[i].charAt(j) == '0') {
                    q.add(new MAP(i, j, 0, 0));
                }
            }
        }

        while (!q.isEmpty()) {
            MAP tmp = q.poll();
            bfs(tmp.x, tmp.y, tmp.key, tmp.cnt, q);
        }

        System.out.println(ans);
        sc.close();
    }
}